domain/keymanager/modelmigration/export.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelmigration
+
+import (
+	"context"
+
+	"github.com/juju/description/v8"
+
+	"github.com/juju/juju/core/modelmigration"
+	"github.com/juju/juju/core/user"
+	accessservice "github.com/juju/juju/domain/access/service"
+	accessstate "github.com/juju/juju/domain/access/state"
+	"github.com/juju/juju/domain/keymanager/service"
+	"github.com/juju/juju/domain/keymanager/state"
+	"github.com/juju/juju/internal/errors"
+)
+
+// exportOperation is the type used to describe the export operation for a
+// model's authorized keys.
+type exportOperation struct {
+	modelmigration.BaseOperation
+
+	service     ExportService
+	userService UserService
+}
+
+// ExportService represents the service methods needed for exporting the
+// authorized keys of a model during migration.
+type ExportService interface {
+	// GetAllUserPublicKeys returns all of the public keys in the model for each
+	// user grouped by [user.UUID].
+	GetAllUsersPublicKeys(context.Context) (map[user.UUID][]string, error)
+}
+
+// Execute the migration of the model's authorized keys.
+func (e *exportOperation) Execute(ctx context.Context, model description.Model) error {
+	usersKeys, err := e.service.GetAllUsersPublicKeys(ctx)
+	if err != nil {
+		return errors.Errorf(
+			"cannot export authorized keys for model while getting all user keys: %w",
+			err,
+		)
+	}
+
+	userIds := make([]user.UUID, 0, len(usersKeys))
+	for userId := range usersKeys {
+		userIds = append(userIds, userId)
+	}
+
+	userIdNameMap, err := e.userService.GetUsernamesForIds(ctx, userIds...)
```

Comment:
So we are making 2 api calls - one for keys, then one for users for the ids from the keys. And we have to account for the awkwardness where an id in a key can't be resolved to a user. But the relational schema enforces this integrity. We should have a bespoke api call to return the keys and corresponding user names directly, composed in one txn. Then we also avoid introducing a bulk call with different semantics.